shell/platform/darwin/ios/framework/Source/FlutterViewController.mm

```diff
@@ -689,8 +690,21 @@ - (void)viewDidLayoutSubviews {
 
   // This must run after updateViewportMetrics so that the surface creation tasks are queued after
   // the viewport metrics update tasks.
-  if (firstViewBoundsUpdate)
+  if (firstViewBoundsUpdate) {
     [self surfaceUpdated:YES];
+
+    flutter::Shell& shell = [_engine.get() shell];
+    fml::TimeDelta waitTime =
+#if NDEBUG
+        fml::TimeDelta::FromMilliseconds(100);
```

Comment:
Are we introducing a new race condition by putting these timeouts here?

Reply:
@tvolkert The current state of affairs is we issue the rendering of a frame asynchronously and we don't pay attention to when it is finished and plow right along on the main thread.  That is what is causing the first frame to be black.  We should be issuing the async render and actually paying attention to when it is done before the main thread continues on.  That's what this is doing.

The reason for the timeout is not because I'm afraid of deadlock, it's because our renderer is inefficient to the point where I'm saying "If rendering the first frame takes over 100ms it's better to show a black frame than to make the main thread wait any longer"

We have outstanding work to do to make the first frame render faster.  Right now we've gated all our initialization logic on knowing the dimensions of the surface we are rendering to, which is arbitrary and inefficient.  In addition to this synchronization we should decouple that initialization that needs dimensions vs initialization that doesn't need dimensions so we can make the first frame even faster.  Right now first frame is ~2.5x slower that the second frame.  Even when we optimize that, we will still need to synchronize.  We can't just fire off an event and not care about when it is finished.

Reply:
I'm curious about the difference between:
1. Showing a black frame at 100ms when this times out
2. Blocking everything until the first frame is rendered (say in 200ms)

To me, they seem equivalent. In both cases, the user won't be able to see any meaningful pixels or interact with the app at 100ms.

Reply:
@liyuqian Both aren't desirable.  Having the timeout is safer since in iOS if you occupy the main thread for too long the operating system will kill your app.  Also, in the case where there is a bug that makes the attempt to request a frame ignored we won't get deadlocked.